File: backend/app/app_tenant/marketplace/schemas/marketplace.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.app_tenant.marketplace.models import MarketplaceAppStatus


class AppCategoryOut(BaseModel):
    id: UUID
    name: str
    slug: str
    sort_order: int

    model_config = {"from_attributes": True}


class MarketplaceAppOut(BaseModel):
    id: UUID
    name: str
    description: str | None
    icon: str | None
    version: str
    status: MarketplaceAppStatus
    is_official: bool
    install_count: int
    category_id: UUID | None
    category_name: str | None = None
    installed: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}


class MarketplaceAppDetail(MarketplaceAppOut):
    manifest: dict


class MarketplaceAppCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    description: str | None = None
    icon: str | None = "📦"
    version: str = "1.0.0"
    category_slug: str | None = None
    manifest: dict = Field(default_factory=dict)
    status: MarketplaceAppStatus = MarketplaceAppStatus.DRAFT


class MarketplaceAppUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=128)
    description: str | None = None
    icon: str | None = None
    version: str | None = None
    category_slug: str | None = None
    manifest: dict | None = None
    status: MarketplaceAppStatus | None = None


class MarketplaceAppCreateFromResources(BaseModel):
    """从当前租户已有资源打包上架（草稿）。"""

    name: str = Field(..., min_length=1, max_length=128)
    description: str | None = None
    icon: str | None = "📦"
    category_slug: str | None = "rag"
    flow_id: UUID | None = None
    agent_id: UUID | None = None
    kb_id: UUID | None = None


class AppInstallOut(BaseModel):
    id: UUID
    app_id: UUID
    app_name: str
    tenant_id: UUID
    flow_id: UUID | None
    agent_id: UUID | None
    kb_id: UUID | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AppInstallResult(BaseModel):
    install: AppInstallOut
    flow_id: UUID | None = None
    agent_id: UUID | None = None
    kb_id: UUID | None = None
    message: str = "安装成功"

File: ui/workbench/lib/types/marketplace.ts
/** 域类型：与 backend OpenAPI 对齐。 */

import type { EnumOption } from "@/lib/enum-meta";
import type { TagRef } from "./tags";
export interface AppCategory {
  id: string;
  name: string;
  slug: string;
  sort_order: number;
}


export interface MarketplaceApp {
  id: string;
  name: string;
  description?: string | null;
  icon?: string | null;
  version: string;
  status: string;
  is_official: boolean;
  install_count: number;
  rating_avg: number;
  rating_count: number;
  visibility?: string;
  category_id?: string | null;
  category_name?: string | null;
  tags?: TagRef[];
  installed: boolean;
  review_note?: string | null;
  submitted_at?: string | null;
  reviewed_at?: string | null;
  created_at: string;
}


export interface AppRating {
  id: string;
  app_id: string;
  user_id: string;
  score: number;
  comment?: string | null;
  created_at: string;
  updated_at: string;
}


export interface MarketplaceAppDetail extends MarketplaceApp {
  manifest: Record<string, unknown>;
  my_rating?: AppRating | null;
}


export interface AppInstallResult {
  install: {
    id: string;
    app_id: string;
    app_name: string;
    flow_id?: string | null;
    agent_id?: string | null;
    kb_id?: string | null;
  };
  flow_id?: string | null;
  agent_id?: string | null;
  kb_id?: string | null;
  message: string;
}


export interface AppUpgradeResult {
  install: AppInstall;
  previous_version: string;
  new_version: string;
  message: string;
}


export interface UpgradeFieldChange {
  field: string;
  label: string;
  before?: string | null;
  after?: string | null;
  changed: boolean;
}


export interface UpgradeResourceDiff {
  resource_type: string;
  resource_id?: string | null;
  resource_name: string;
  changes: UpgradeFieldChange[];
  has_changes: boolean;
}


export interface AppUpgradePreview {
  app_id: string;
  app_name: string;
  installed_version: string;
  target_version: string;
  can_upgrade: boolean;
  has_changes: boolean;
  message?: string | null;
  resources: UpgradeResourceDiff[];
}


export interface AppRollbackPreview {
  app_id: string;
  app_name: string;
  current_version: string;
  target_version: string;
  can_rollback: boolean;
  message?: string;
  resources: UpgradeResourceDiff[];
}


export interface AppInstall {
  id: string;
  app_id: string;
  app_name: string;
  installed_version?: string;
  app_version?: string | null;
  flow_id?: string | null;
  agent_id?: string | null;
  kb_id?: string | null;
  created_at: string;
}
